tss-esapi/tests/integration_tests/structures_tests/tagged_tests/tagged_signature_scheme_tests.rs

```diff
@@ -0,0 +1,138 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use tss_esapi::{
+    interface_types::algorithm::{HashingAlgorithm, SignatureSchemeAlgorithm},
+    structures::{EcDaaScheme, HashScheme, HmacScheme, SignatureScheme},
+    tss2_esys::{
+        TPMS_SCHEME_ECDAA, TPMS_SCHEME_HASH, TPMS_SCHEME_HMAC, TPMT_SIG_SCHEME, TPMU_SIG_SCHEME,
+    },
+};
+
+use std::convert::{TryFrom, TryInto};
+
+fn validate_tss_hash_scheme(
+    left: &TPMS_SCHEME_HASH,
+    right: &TPMS_SCHEME_HASH,
+    union_field_name: &str,
+) {
+    assert_eq!(
+        left.hashAlg, right.hashAlg,
+        "{} in details, hashAlg did not match",
+        union_field_name
+    );
+}
+
+fn validate_tss_ecdaa_scheme(
+    left: &TPMS_SCHEME_ECDAA,
+    right: &TPMS_SCHEME_ECDAA,
+    union_field_name: &str,
+) {
+    assert_eq!(
+        left.hashAlg, right.hashAlg,
+        "{} in details, hashAlg did not match",
+        union_field_name
+    );
+    assert_eq!(
+        left.count, right.count,
+        "{} in details, count did not match",
+        union_field_name
+    );
+}
+
+fn validate_tss_hmac_scheme(
+    left: &TPMS_SCHEME_HMAC,
+    right: &TPMS_SCHEME_HMAC,
+    union_field_name: &str,
+) {
+    assert_eq!(
+        left.hashAlg, right.hashAlg,
+        "{} in details, hashAlg did not match",
+        union_field_name
+    );
+}
+
+macro_rules! test_valid_conversions {
+    (SignatureScheme::$item:ident, $union_field_name:ident, $tss_details_field_validator:expr, $native_scheme_field:ident, $native_scheme:expr) => {
+        let tss_actual: TPMT_SIG_SCHEME = SignatureScheme::$item { $native_scheme_field: $native_scheme }
+        .try_into()
+        .expect(&format!("Failed to convert {} signature scheem into TSS type", stringify!($item)));
+
+
+        let tss_expected = TPMT_SIG_SCHEME {
+            scheme: SignatureSchemeAlgorithm::$item.into(),
+            details: TPMU_SIG_SCHEME {
+                $union_field_name: $native_scheme.into(),
+            },
+        };
+
+        assert_eq!(
+            tss_actual.scheme, tss_expected.scheme,
+            "scheme for Actual converted value did not match expected in TSS types for SignatureScheme {}", stringify!($item),
+        );
+
+        $tss_details_field_validator(
+            &unsafe { tss_actual.details.$union_field_name },
+            &unsafe { tss_expected.details.$union_field_name },
+            stringify!($union_field_name),
+        );
+
+        let native_expected = SignatureScheme::$item { $native_scheme_field: $native_scheme };
+
+        let native_actual = SignatureScheme::try_from(tss_expected)
+            .expect(&format!("Failed to convert TSS type into {}", stringify!($item)));
+
+        assert_eq!(
+            native_actual, native_expected,
+            "The actual SignatureScheme did not match expected",
+        );
+    };
+}
+
+macro_rules! test_valid_conversions_hash_scheme {
+    (SignatureScheme::$item:ident, $union_field_name:ident) => {
```

Comment:
Hmm, I think you might be able to keep all of the cases in the same macro with something like
```
(SignatureScheme::EcDaa, $union_field_name:ident) => {
```
before the generic macro, but I'm not entirely sure

Reply:
Interesting, I will look into that. Thanks.

Reply:
Fixed